fdmi/plugins/fdmi_sample_plugin.c

```diff
@@ -0,0 +1,426 @@
+			if (len > 0) {
+				m0_console_printf("\"cr_val\": \"%s\"",
+						  to_hex(cr_rec[i].cr_val.u.ab_buf.b_addr, len));
+			} else {
+				m0_console_printf("\"cr_val\": \"0\"");
+			}
+			m0_console_printf(" }\n");
+
+		}
+	} m0_tl_endfor;
+}
+
+static void fsp_usage(void)
+{
+	fprintf(stderr,
+		"Usage: fdmi_sample_plugin "
+		"-l local_addr -h ha_addr -p profile_fid -f process_fid "
+		"-g fdmi_plugin_fid\n"
+		"Use -? or -i for more verbose help on common arguments.\n"
+		"Usage example for common arguments: \n"
+		"fdmi_sample_pluginm0sched -l 192.168.52.53@tcp:12345:4:1 "
+		"-h 192.168.52.53@tcp:12345:1:1 "
+		"-p 0x7000000000000001:0x37 -f 0x7200000000000001:0x19"
+		"-g 0x6c00000000000001:0x51"
+		"\n");
+}
+
+/**
+ * @retval 0      Success.
+ * @retval -Exxx  Error.
+ */
+static int fsp_args_parse(struct m0_fsp_params *params, int argc, char ** argv)
+{
+	int rc = 0;
+
+	params->spp_local_addr 	= NULL;
+	params->spp_hare_addr   = NULL;
+	params->spp_profile_fid = NULL;
+	params->spp_process_fid = NULL;
+	params->spp_fdmi_plugin_fid_s = NULL;
+
+	rc = M0_GETOPTS("fdmi_sample_plugin", argc, argv,
+			M0_HELPARG('?'),
+			M0_VOIDARG('i', "more verbose help",
+				   LAMBDA(void, (void) {
+						   fsp_usage();
+						   exit(0);
+					   })),
+			M0_STRINGARG('l', "Local endpoint address",
+				     LAMBDA(void, (const char *string) {
+						     params->spp_local_addr = (char*)string;
+					     })),
+			M0_STRINGARG('h', "HA address",
+				     LAMBDA(void, (const char *str) {
+						     params->spp_hare_addr = (char*)str;
+					     })),
+			M0_STRINGARG('f', "Process FID",
+				     LAMBDA(void, (const char *str) {
+						     params->spp_process_fid = (char*)str;
+					     })),
+			M0_STRINGARG('p', "Profile options for client",
+				     LAMBDA(void, (const char *str) {
+						     params->spp_profile_fid = (char*)str;
+					     })),
+			M0_STRINGARG('g', "FDMI plugin fid",
+				     LAMBDA(void, (const char *str) {
+						     params->spp_fdmi_plugin_fid_s =
+							     (char*)str;
+					     })));
+	if (rc != 0)
+		return M0_ERR(rc);
+
+        /* All mandatory params must be defined. */
+	if (rc == 0 &&
+	    (params->spp_local_addr == NULL || params->spp_hare_addr == NULL ||
+	     params->spp_profile_fid == NULL || params->spp_process_fid == NULL ||
+	     params->spp_fdmi_plugin_fid_s == NULL)) {
+		fsp_usage();
+		return M0_ERR(-EINVAL);
+	}
+
+	rc = m0_fid_sscanf(params->spp_fdmi_plugin_fid_s, &params->spp_fdmi_plugin_fid);
+	if (rc != 0) {
+		rc = M0_ERR_INFO(rc, "Invalid FDMI plugin fid format: fid=%s",
+				 params->spp_fdmi_plugin_fid_s);
+	}
+
+	return rc;
+}
+
+/**
+ * Callback function called by the FDMI service to deliver the FDMI event
+ * to the plugin program.
+ */
+static int process_fdmi_record(struct m0_uint128 *rec_id,
+			       struct m0_buf fdmi_rec,
+		   	       struct m0_fid filter_id)
+{
+	struct m0_fol_rec fol_rec;
+
+	m0_fol_rec_init(&fol_rec, NULL);
+	m0_fol_rec_decode(&fol_rec, &fdmi_rec);
```

Comment:
Please check the result of this function and handle error.

Reply:
done